modify HTMLHtmlElement to reject async-hide when set by JS

snyderp opened the pull request:
## Submitter Checklist:

- [X] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [X] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [X] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests && npm run test-security`) on
  - [ ] Windows
  - [X] macOS
  - [ ] Linux
- [X] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection

## Test Plan:


## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

petemill commented on patches/third_party-blink-renderer-core-html-html_html_element.cc.patch:
```diff
@@ -0,0 +1,41 @@
+ }
+ 
++void HTMLHtmlElement::AttributeChanged(const AttributeModificationParams& params) {
++  // Check to see if the following conditions are true, to work around
```
This is a large patch. Is there a way we can patch in a hook to a more generic `AttributeChanged` brave function, either in chromium_src, or brave?

snyderp replied:
@petemill sure, this could be revised to pass a `AttributeModificationParams` into some func (`PossiblyModifyAttributeModificationParamsToIgnoreAsyncHide`?) which could pass back the same, or a modified `AttributeModificationParams` as necessary.

I dont have much experience with that pattern at Brave though (all my work has been 100% in blink).  Who would be good to ask for some handholding there :)

petemill replied:
I think the simplest thing we could do here is to patch in a generic function: `PossiblyModifyAttributeModificationParams`. And then define that function in `brave/chromium_src/third_party/blink/renderer/core/html/html_html_element.cc`. That file could look something like:

```cpp
// define function for `PossiblyModifyAttriguteModificationParams`
AttributeModificationParams PossiblyModifyAttriguteModificationParams(const AttributeModificationParams& original_params) {
  // ...decide to do something or not...
}

// include original chromium blink source which is patched to call above function
#include "../../../../../..//third_party/blink/renderer/core/html/html_html_element.cc"
```

What this buys us is that the patch is minimized, the code is in a .cc file and not a .patch file, and when we want to add more modifications, we might not need to do any further patching. Seems like a minimal improvement, but I think that's the pattern we're going for right now.

snyderp commented:
@petemill Okie dokie, this should be ready for another review now

(I also rebased / squashed to keep the change up to date, but seems like I added a lot of noise above.  Is there another way I should have done this?)

petemill commented on patches/third_party-blink-renderer-core-dom-element.cc.patch:
```diff
@@ -0,0 +1,45 @@
+   }
+ 
++  // Brave addition
++  AtomicString modified_value = PossiblyModifyAttrParam(*this, name, new_value);
```
Can we reduce the patch further by just re-assigning `new_value` here? Is that possible without affecting callers?

bridiver commented on patches/third_party-blink-renderer-core-dom-element.cc.patch:
```diff
@@ -0,0 +1,45 @@
+ #include "third_party/blink/renderer/platform/wtf/text/string_builder.h"
+ #include "third_party/blink/renderer/platform/wtf/text/text_position.h"
+ 
++#if defined(BRAVE_CHROMIUM_BUILD)
```
this kind of stuff should be added through chromium_src override, but I think there is probably a much easier way to do this that doesn't require patching. Investigating now

bridiver commented:
this seems like something that would be better handled through a content script that stubs out whatever it is looking for. I'm also concerned that we're applying this to every single site and also anything we add in the pipeline here is a potential performance issue

bridiver commented:
we can also handle this through a site script that returns a stub library in place of the blocked resource

snyderp commented:
@bridiver the reason to apply to all sites is that we dont have any kind of list of which sites its needed for, and the expectation is "many".

We could do through JS injection too, though we'll need to build stubs for all google properties that use this "trick".

The tricky part here is that the problematic JS is inline, but the blocked resource is remote, and we lack a way of blocking the inline JS.  If JS injection is preferred, that makes sense to me, but there are at least cases where it'll fail that this approach succeeds (though, maybe the simplicity of a content script, along with skipping an "if" statement on each attribute set is worth it, i coulnd't say)

snyderp commented:
closing this in favor of https://github.com/brave/brave-core/pull/2488
